lib/core/src/display/symbol/mesh.rs

```diff
@@ -86,163 +146,106 @@ impl<OnDirty:Callback0+Clone> Mesh<OnDirty> {
-        out_color = vec4(local, 1.0);
-        out_color = _position;
-        out_color = vec4(1.0,1.0,1.0,1.0);
+    /// Creates a new VertexArrayObject, discovers all variable bindings from material to geometry,
+    /// and initializes the VAO with the bindings.
+    fn init_vao(&mut self) {
+        self.vao = Some(VAO::new(&self.context));
+        self.with_program(|program|{
+            let var_bindings = self.discover_variable_bindings();
+            for (variable,opt_scope_type) in &var_bindings {
+                if let Some(scope_type) = opt_scope_type {
+                    let opt_scope = self.geometry.var_scope(*scope_type);
+                    match opt_scope {
+                        None => self.logger.error("Internal error. Invalid var scope."),
+                        Some(scope) => {
+                            let vtx_name = shader::builder::mk_vertex_name(&variable);
+                            let location = self.context.get_attrib_location(program, &vtx_name);
+                            if location < 0 {
+                                self.logger.error(|| format!("Attribute '{}' not found.",vtx_name));
+                            } else {
+                                let location     = location as u32;
+                                let buffer       = &scope.buffer(&variable).unwrap();
+                                let is_instanced = scope_type == &geometry::ScopeType::Instance;
+                                buffer.bind(webgl::Context::ARRAY_BUFFER);
+                                buffer.vertex_attrib_pointer(location, is_instanced);
+                            }
+                        }
+                    }
+                }
+            }
+        });
     }
-"#,
-            )
-                .unwrap();
-
-//            println!("{:?}", self.context.get_shader_info_log(&vert_shader));
-//            println!("{:?}", self.context.get_shader_info_log(&frag_shader));
-
-
-            let program =
-                webgl::link_program(&self.context, &vert_shader, &frag_shader).unwrap();
-
-
-            // === Rendering ==
 
-            self.context.use_program(Some(&program));
+    /// For each variable from the material definition, looks up its position in geometry scopes.
+    pub fn discover_variable_bindings(&self) -> Vec<(String,Option<geometry::ScopeType>)> {
+        let variables = self.material.collect_variables();
+        variables.into_iter().map(|variable| {
+            let target = self.geometry.lookup_variable(&variable);
+            if target.is_none() {
+                let msg = || format!("Unable to bind variable '{}' to geometry buffer.", variable);
+                self.logger.warning(msg);
+            }
+            (variable,target)
+        }).collect()
+    }
 
+    /// Runs the provided function in a context of active program and active VAO. After the function
+    /// is executed, both program and VAO are bound to None.
```

Comment:
Why we do the program and VAO unbinding? To avoid bugs?

Reply:
Yes. I don't want to debug for months a program when I dont unbind VAO and other code (like text) accidentally binds me some more parameters to the VAO. If you are concerned about performance you should not, as this is rather small op. In case we really like to cut it out because of fine tunning performance to the maximum, we can cut it off in production build.